### DISCUSSION TITLE
Create and edit bookmarks

### DISCUSSION THREAD

**Discussion by abelokoj-github** [sumatrapdfreader/sumatrapdf#3044] [Ideas]
I will like an option to be able to create and edit Bookmarks on pdf just like you can do in Foxit pdf reader (The free version). 

In Sumatra pdf reader, there is an option with the shortcut **CTRL+B** that allows you to add a page to favorites. I think there should also be a shortcut say **CRTL+SHFT+B** to add the page to bookmarks and give it a name of your choice.

Users should also be able to click on a pdf bookmark and get the option to rename, delete or change the bookmark order in any pdf document.

May thanks in anticipation. Your work and efforts is highly appreciated.

**Comment by shikov** [comment]
I vote for that. Bookmark editing is more important than annotations.

**Reply by GitHubRulesOK** [reply]
bookmark editing requires full acrobat PRO or similar document level editing/rewrite of contents, easily done by a command line library but MuTool does not yet offer such funtion, even if PyMuPDF might https://stackoverflow.com/questions/69985007/add-bookmarks-to-pdf-using-pymupdf
 annotations are more easily added by readers as done in MuPDF-GL

LATER EDIT (2026)
Bin-2 has recently extended MuPDF abilities to adjust bookmarks see below.

**Reply by shikov** [reply]
I think a new API has been added to the latest version of MuPDF library (1.20) for outline editing (https://mupdf.com/releases/history.html)

**Reply by abelokoj-github** [reply]
I think this will work since I guess Sumatra uses Mupdf

**Reply by abelokoj-github** [reply]
**There are simpler way to implement this without requiring full Adobe pro heavy-kind of software.**

There is an awesome **cross platform light weight pdf tool** known as **coherrent pdf `cpdf`** that allows you do edit bookmarks directly from the command line. It also has a lot of other functions which are very useful. 

I believe Sumatra pdf can simply create ui elements that call cpdf on  background from the command line to perform the bookmark editing functions as well as other great features tgat are supported by `cpdf`. I believe a lot of users will greatly beenefit from this integration. 

**URL:** [https://www.coherentpdf.com/](https://www.coherentpdf.com/)

![Screenshot_20260618_215522_Opera.jpg](https://github.com/user-attachments/assets/0b711f38-81a4-41ad-a347-a2508a2d80b1)

**Comment by meteorquake** [comment]
I find the bookmark editing in the free FoxitReader superbly useful, and Sumatra conveniently adds a menu item for Foxit. When you save in Foxit, Sumatra automatically updates to show the bookmarks (or manually if the file is large).
So in a sense, for me the bookmark function is already there using the menu File - Open in FoxitReader, and it is very smooth.
If Bookmarks were added as a direct feature that might be a bonus, but for those needing it, the above outlines the form it is currently there in.
d

**Comment by meteorquake** [comment]
NB FoxitReader seems to have removed bookmark adding/editing with the latest update. If anyone uses that feature, best not to update, or do a roll-back!

**Reply by GitHubRulesOK** [reply]
@meteorquake 
AFAIK my prefered sub editor :-) Tracker Exchange  XPDFedit still does bookmarks in the Free version I use, but you need to avoid some others that may be free in Foxit or else pay to remove demo Tracker watermarks
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/974a5898-66d7-46e6-ae9c-978409738165)
![image](https://github.com/sumatrapdfreader/sumatrapdf/assets/12754913/9d4a42f2-c38a-4296-8f87-10b8a4230738)

**Comment by Bin-2** [comment]


**Reply by GitHubRulesOK** [reply]
@bin-2 good PR **are always welcome especially if FOCUSED on a BIG benefit** with small changes the most likely reason for not included is too many changes to review and accept as one ROBUST change.

Recent exposure of MuTool internals has meant a lot of quick modifications but higher amount of collateral damage when not "soak-tested".

Docking Annotations is nice as is editing the PDF with annotations or similar goto's like BookMarks but it does mean the file is either APPENDED or ALTERED and that must not affect the documents integrity.

**Reply by GitHubRulesOK** [reply]
@Bin-2 my comment was not intended to discourage simply to warn that changing a file without a novice user getting a warning about overwrite could be a consideration.
E.G. a lot of users complained their files became unusable when annotations modified the files against expectation. and since bookmarking has to do similar any FINAL overwrite file should include confirmation from user.

**Reply by GitHubRulesOK** [reply]
as to encoded files with passwords or signatures the easiest way is remove all such waste of time so they can be replaced if anyone wants shorter easy lifetime. I agree this is not that different to go to annotations or other annotations but because an edit (same as remove DRM or signatures) it changes a file and that could be VERY easily done without confirmation/warning here
<img width="277" height="152" alt="image" src="https://github.com/user-attachments/assets/156275d9-a09d-4492-a029-873d55b528fc" />

What I am suggesting is the bookmarks editing is good but then only needs to hook into existing file saving without any user interface changes. (Keep the PR simple and focused on the bookmarks panel)

**Reply by meteorquake** [reply]
Regarding the comment about Foxit Reader no longer editing bookmarks, you just have installed a slightly older version that does and switch off the updater. I have 12.1.3.15356 installed. You'll have to pull the installer off the web somewhere but if you want to check validity you can check the MD5 hash on a virus scanner website to see its age and validity. Usually I'll also check the MD5s from several independent reliable sites and check their MD5s match (if necessary, download and compute MD5s of several).

**Reply by abelokoj-github** [reply]
@meteorquake 


**There are simpler way to implement this without requiring full Adobe pro heavy-kind of software.**

There is an awesome **cross platform light weight pdf tool** known as **coherrent pdf `cpdf`** that allows you do edit bookmarks directly from the command line. It also has a lot of other functions which are very useful. 

I believe Sumatra pdf can simply create ui elements that call cpdf on  background from the command line to perform the bookmark editing functions as well as other great features tgat are supported by `cpdf`. I believe a lot of users will greatly beenefit from this integration. 

**URL:** [https://www.coherentpdf.com/](https://www.coherentpdf.com/)

![Screenshot_20260618_215522_Opera.jpg](https://github.com/user-attachments/assets/0b711f38-81a4-41ad-a347-a2508a2d80b1)

**Comment by GitHubRulesOK** [comment]
@abelokoj-github I often use cpdf for scripts since as you say it has features way beyond most similar tools but bookmarking is via command lines outside SumatraPDF (which now uses Mutool scripting) and bookmarks is a big task outside of a pdf GUI editor.
Hence by far the simplest is use the free function in Tracker PDFXedit. But sticking to SumatraPDF it has been shown to be possible internally.
<img width="535" height="293" alt="image" src="https://github.com/user-attachments/assets/ecbf8333-ed4d-45dd-9eea-7e7e38cb4e4e" />
